File: pycppad/adfun.py
# $begin independent$$ $newlinech #$$
# $spell
#	Numpy
#	tuple
# $$
#
# $section Create an Independent Variable Vector$$
#
# $head Syntax$$
# $icode%a_x% = independent(%x%)%$$
#
# $index independent, variables$$
# $index variables, independent$$
# $index recording, start$$
# $index start, recording$$
#
# $head Purpose$$
# Creates an independent variable vector and starts recording operations
# involving objects that are instances of $codei%type(%a_x%[0])%$$.
# You must create an $cref/adfun/$$ object, or use $cref/abort_recording/$$,
# to stop the recording before making another call to $code independent$$,
#
# $head x$$ 
# The argument $icode x$$ must be a $code numpy.array$$ with one dimension
# (i.e., a vector).
# All the elements of $icode x$$ must all be of the same type and
# instances of either $code int$$, $code float$$ or $code  a_float$$.
#
# $head a_x$$ 
# The return value $icode a_x$$ is a $code numpy.array$$ 
# with the same shape as $icode x$$. 
# If the elements of $icode x$$ are instances of $code int$$ or $code float$$
# the elements of $icode a_x$$ are instances of $code a_float$$.
# If the elements of $icode x$$ are instances of $code a_float$$
# the elements of $icode a_x$$ are instances of $code a2float$$.
# The $cref/value/$$ of the elements of $icode a_x$$ 
# are equal to the corresponding elements of $icode x$$.
# 
# $children%
#	example/independent.py
# %$$
# $head Example$$
# The file $cref/independent.py/$$ 
# contains an example and test of this function.
#
# $end
# ---------------------------------------------------------------------------
# $begin adfun$$ $newlinech #$$
# $spell
#	numpy
#	len
#	adfun
# $$
#
# $section Create an AD Function Object$$
#
# $index dependent, variables$$
# $index variables, dependent$$
# $index recording, stop$$
# $index stop, recording$$
#
# $head Syntax$$
# $icode%f% = adfun(%a_x%, %a_y%)%$$
#
# $head Purpose$$
# The function object $icode f$$ will store the $codei%type( %a_x%[0] )%$$
# operation sequence that mapped the independent variable vector 
# $icode a_x$$ to the dependent variable vector $icode a_y$$.
#
# $head a_x$$
# The argument $icode a_x$$ is the $code numpy.array$$ 
# returned by the previous call to $cref/independent/$$.
# Neither the size of $icode a_x$$, or the value it its elements,
# may change between calling
# $codei%
#	%a_x% = independent(%x%)
# %$$
# and
# $codei%
#	%f% = adfun(%a_x%, %a_y%)
# %$$
# The length of the vector $icode a_x$$ determines the domain size
# $latex n$$ for the function $latex y = F(x)$$ below.
#
# $head a_y$$
# The argument $icode a_y$$ specifies the dependent variables.
# It must be a $code numpy.array$$ with one dimension
# (i.e., a vector) and with the same type of elements as $icode a_x$$.
# The object $icode f$$ stores the $codei%type( %a_x%[0] )%$$ operations 
# that mapped the vector $icode a_x$$ to the vector $icode a_y$$.
# The length of the vector $icode a_y$$ determines the range size
# $latex m$$ for the function $latex y = F(x)$$ below.
#
# $head f$$
# The return value $icode f$$ can be used to evaluate the function
# $latex \[
#	F : \B{R}^n \rightarrow \B{R}^m
# \] $$
# and its derivatives, where $latex y = F(x)$$ corresponds to the 
# operation sequence mentioned above.
#
# $subhead m$$
# The range size $latex m$$ is equal to the length of the vector $icode a_y$$.
#
# $subhead n$$
# The domain size $latex n$$ is equal to the length of the vector $icode a_x$$.
#
# $subhead level$$
# The $cref/ad/$$ level for the object $icode f$$ is one less than
# the AD level for the arguments $icode a_x$$ and $icode a_y$$;
# i.e., if $codei%type( %a_x%[0] )%$$ is $code a_float$$ ($code a2float$$)
# the corresponding AD level for $icode f$$ is zero (one).
#
# $children%
#	example/adfun.py
# %$$
# $head Example$$
# The file $cref/adfun.py/$$ 
# contains an example and test of this function.
#
# $end
# ---------------------------------------------------------------------------
# $begin abort_recording$$ $newlinech #$$
# $comment implemented in pycppad.cpp$$
# $spell 
# $$
# 
# $section Abort a Recording of AD Operations$$
# 
# $index abort, AD recording$$
# $index AD, abort recording$$
# $index recording, abort AD$$
# $index independent, abort recording$$
# 
# $head Syntax$$
# $codei%abort_recording()%$$
#
#$head Purpose$$
# Sometimes it is necessary to abort the recording of AD operations
# that started with a call of the form
# $codei%
# 	%a_x% = independent(%x%)
# %$$
# If such a recording is currently in progress,
# this will stop the recording and delete the corresponding information.
# Otherwise, $code abort_recording$$ has no effect.
#
# $children%
#	example/abort_recording.py
# %$$
# $head Example$$
# The file $cref/abort_recording.py/$$ 
# contains an example and test of this operation.
# It returns true if it succeeds and false otherwise.
#
# $end
# ---------------------------------------------------------------------------
import cppad_
import numpy
 
def independent(x) :
  """
  a_x = independent(x): create independent variable vector a_x, equal to x,
  and start recording operations that use the class corresponding to ad( x[0] ).
  """
  #
  # It would be better faster if all this type checking were done in the C++
  #
  if not isinstance(x, numpy.ndarray) :
    raise NotImplementedError('independent(x): x is not of type numpy.array')
  #
  x0 = x[0]
  if isinstance(x0, int) :
    for j in range( len(x) ) :
      if not isinstance(x[j], int) :
        other = 'x[' + str(j) + '] is ' + type(x[j]).__name__
        msg   = 'independent(x): mixed types x[0] is int and ' + other
        raise NotImplementedError(msg)
    x = numpy.array(x, dtype=int)       # incase dtype of x is object
    return cppad_.independent(x, 1)     # level = 1
  #
  if isinstance(x0, float) :
    for j in range( len(x) ) :
      if not isinstance(x[j], float) :
        other = 'x[' + str(j) + '] is ' + type(x[j]).__name__
        msg   = 'independent(x): mixed types x[0] is float and ' + other
        raise NotImplementedError(msg)
    x = numpy.array(x, dtype=float)     # incase dtype of x is object
    return cppad_.independent(x, 1)     # level = 1
  #
  if isinstance(x0, cppad_.a_float) :
    for j in range( len(x) ) :
      if not isinstance(x[j], cppad_.a_float) :
        other = 'x[' + str(j) + '] is ' + type(x[j]).__name__
        msg   = 'independent(x): mixed types x[0] is a_float and ' + other
        raise NotImplementedError(msg)
    return cppad_.independent(x, 2)     # level = 2
  #
  msg = 'independent(x): x[0] has type' + type(x0).__name__ + '\n'
  msg = 'only implemented where x[j] is int, float, or a_float'
  raise NotImplementedError(msg)

class adfun_float(cppad_.adfun_float) :
  """
  Create a level zero function object (evaluates using floats).
  """
  # Kludge: The following reshaping should be done in adfun.cpp
  def jacobian(self, x) :
    return self.jacobian_(x).reshape(self.range(), self.domain())
  def hessian(self, x, w) :
    return self.hessian_(x, w).reshape(self.domain(), self.domain())
  pass

class adfun_a_float(cppad_.adfun_a_float) :
  """
  Create a level one function object (evaluates using a_float).
  """
  # Kludge: The following reshaping should be done in adfun.cpp
  def jacobian(self, x) :
    return self.jacobian_(x).reshape(self.range(), self.domain())
  def hessian(self, x, w) :
    return self.hessian_(x, w).reshape(self.domain(), self.domain())
  pass

def adfun(x,y) :
  """
  f = adfun(x,y): Stop recording and place it in the function object f.
  x: a numpy one dimnesional array containing the independent variable vector.
  y: a vector with same type as x and containing the dependent variable vector.
  """
  #
  # It would be better faster if all this type checking were done in the C++
  #
  if not isinstance(x, numpy.ndarray) or not isinstance(y, numpy.ndarray) :
    raise NotImplementedError('adfun(x, y): x or y is not of type numpy.array')
  #
  x0 = x[0]
  y0 = y[0]
  if isinstance(x0, cppad_.a_float) :
    for j in range( len(x) ) :
      if not isinstance(x[j], cppad_.a_float) :
        other = 'x[' + str(j) + '] is ' + type(x[j]).__name__
        msg   = 'adfun(x, y): mixed types x[0] is a_float and ' + other
        raise NotImplementedError(msg)
    #
    for i in range( len(y) ) :
      if not isinstance(y[i], cppad_.a_float) :
        other = 'y[' + str(i) + '] is ' + type(y[i]).__name__
        msg   = 'adfun(x, y): mixed types x[0] is a_float and ' + other
        raise NotImplementedError(msg)
    #
    return adfun_float(x, y)
  #
  if isinstance(x0, cppad_.a2float) :
    for j in range( len(x) ) :
      if not isinstance(x[j], cppad_.a2float) :
        other = 'x[' + str(j) + '] is ' + type(x[j]).__name__
        msg   = 'independent(x): mixed types x[0] is a2float and ' + other
        raise NotImplementedError(msg)
    #
    for i in range( len(y) ) :
      if not isinstance(y[i], cppad_.a2float) :
        other = 'y[' + str(i) + '] is ' + type(y[i]).__name__
        msg   = 'independent(x): mixed types x[0] is a2float and ' + other
        raise NotImplementedError(msg)
    #
    return adfun_a_float(x, y)
  #
  raise NotImplementedError(
        'adfun(x, y): elements of x and y are not a_float or a2float')

File: pycppad/adfun.cpp
/*
---------------------------------------------------------------------------
$begin forward$$
$spell
	numpy
	adfun
	Taylor
$$

$section  Forward Mode: Derivative in One Domain Direction$$

$index forward$$
$index domain, direction derivative$$
$index derivative, domain direction$$
$index direction, domain derivative$$


$head Syntax$$
$icode%y_p% = %f%.forward(%p%, %x_p%)%$$

$head Purpose$$
We use $latex F : \B{R}^n \rightarrow \B{R}^m$$ to denote the 
function corresponding to the $code adfun$$ object $cref/f/adfun/f/$$.
Given the $th p$$ order Taylor expansion for a function
$latex X : \B{R} \rightarrow \B{R}^n$$, this function can be used
to compute the $th p$$ order Taylor expansion for the function
$latex Y : \B{R} \rightarrow \B{R}^m$$ defined by
$latex \[
	Y(t) = F [ X(t) ]
\] $$

$head x_k$$
For $latex k = 0 , \ldots , p$$,
we use $latex x^{(k)}$$ to denote the value of $icode x_k$$ in the
most recent call to
$codei%
	%f%.forward(%k%, %x_k%)
%$$
including $latex x^{(p)}$$ as the value $icode x_p$$ in this call.
We define the function $latex X(t)$$ by
$latex \[
	X(t) =  x^{(0)} + x^{(1)} * t + \cdots + x^{(p)} * t^p 
\] $$

$head y_k$$
For $latex k = 0 , \ldots , p$$,
we use $latex y^{(k)}$$ to denote the Taylor coefficients
for $latex Y(t) = F[ X(t) ]$$ expanded about zero; i.e.,
$latex \[
\begin{array}{rcl}
y^{(k)} & = & Y^{(k)} (0) / k !
\\
Y(t)    & = & y^{(0)} + y^{(1)} * t + \cdots + y^{(p)} * t^p + o( t^p )
\end{array}
\] $$
where $latex o( t^p ) / t^p \rightarrow 0$$ as $latex t \rightarrow 0$$.
The coefficient $latex y^{(p)}$$ is equal to 
the value $icode y_p$$ returned by this call.

$head f$$
The object $icode f$$ must be an $cref adfun$$ object.
We use $cref/level/adfun/f/level/$$ for the AD $cref ad$$ level of 
this object.

$head p$$
The argument $icode p$$ is a non-negative $code int$$.
It specifies the order of the Taylor coefficient for $latex Y(t)$$
that is computed.

$head x_p$$
The argument $icode x_p$$ is a $code numpy.array$$ with one dimension
(i.e., a vector) with length equal to the domain size $cref/n/adfun/f/n/$$
for the function $icode f$$.
It specifies the $th p$$ order Taylor coefficient for $latex X(t)$$.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is zero,
all the elements of $icode x_p$$ must be either $code int$$ or instances
of $code float$$.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is one,
all the elements of $icode x_p$$ must be $code a_float$$ objects.


$head y_p$$
The return value $icode y_p$$ is a $code numpy.array$$ with one dimension
(i.e., a vector) with length equal to the range size $cref/m/adfun/f/m/$$
for the function $icode f$$.
It is set to the $th p$$ order Taylor coefficient for $latex Y(t)$$.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is zero,
all the elements of $icode y_p$$ will be instances of $code float$$.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is one,
all the elements of $icode y_p$$ will be $code a_float$$ objects.

$children%
	example/forward_0.py%
	example/forward_1.py
%$$
$head Example$$
$table
$rref forward_0.py$$
$rref forward_1.py$$
$tend

$end
---------------------------------------------------------------------------
$begin reverse$$
$spell
	dw
	numpy
	adfun
	Taylor
$$

$section  Reverse Mode: Derivative in One Range Direction$$

$index reverse$$
$index derivative, range direction$$
$index range, direction derivative$$
$index direction, range derivative$$

$head Syntax$$
$icode%dw% = %f%.forward(%p%, %w%)%$$

$head Purpose$$
Reverse mode computes the derivative of the $cref forward$$ more 
Taylor coefficients with respect to the domain variable $latex x$$.

$head x_k$$
For $latex k = 0 , \ldots , p$$,
we use $latex x^{(k)}$$ to denote the value of $icode x_k$$ in the
most recent call to
$codei%
	%f%.forward(%k%, %x_k%)
%$$
We use $latex F : \B{R}^n \rightarrow \B{R}^m$$ to denote the 
function corresponding to the $code adfun$$ object $cref/f/adfun/f/$$.

$head X(t, u)$$
We define the function $latex X : \B{R} \times \B{R}^n \rightarrow \B{R}^n$$ by
$latex \[
	X(t, u) =  u + x^{(0)} + x^{(1)} * t + \cdots + x^{(p-1)} * t^{p-1} 
\] $$
Note that for $latex k = 0 , \ldots , p - 1$$,
$latex \[
	x^{(k)} = \frac{1}{k !} \frac{\partial^k}{\partial t^k} X(0, 0)
\] $$

$head W(t, u)$$
The function $latex W : \B{R} \times \B{R}^n \rightarrow \B{R}$$
is defined by
$latex \[
W(t, u) = w_0 * F_0 [ X(t, u) ] + \cdots + w_{m-1} * F_{m-1} [ X(t, u) ]
\] $$
We define the function $latex W_k : \B{R}^n \rightarrow \B{R}$$ by
$latex \[
	W_k ( u ) = \frac{1}{k !} \frac{\partial^k}{\partial t^k} W(0, u)
\] $$
It follows that
$latex \[
W(t, u ) = W_0 ( u ) + W_1 ( u ) * t + \cdots + W_{p-1} (u) * t^{p-1}
         + o( t^{p-1} )
\] $$
where $latex o( t^{p-1} ) / t^{p-1} \rightarrow 0$$
as $latex t \rightarrow 0$$.

$head f$$
The object $icode f$$ must be an $cref adfun$$ object.
We use $cref/level/adfun/f/level/$$ for the AD $cref ad$$ level of 
this object.

$head p$$
The argument $icode p$$ is a non-negative $code int$$.
It specifies the order of the Taylor coefficient $latex W_{p-1} ( u )$$
that is differentiated.
Note that $latex W_{p-1} (u)$$ corresponds a derivative of order 
$latex p-1$$ of $latex F(x)$$,
so the derivative of $latex W_{p-1} (u)$$ corresponds to a derivative
of order $latex p$$ of $latex F(x)$$.

$head w$$
The argument $icode w$$ is a $code numpy.array$$ with one dimension
(i.e., a vector) with length equal to the range size $cref/m/adfun/f/m/$$
for the function $icode f$$.
It specifies the weighting vector $latex w$$ used in the definition of
$latex W(t, u)$$.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is zero,
all the elements of $icode w$$ must be either $code int$$ or instances
of $code float$$.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is one,
all the elements of $icode w$$ must be $code a_float$$ objects.

$head dw$$
The return value $icode v$$ is a $code numpy.array$$ with one dimension
(i.e., a vector) with length equal to the domain size $cref/n/adfun/f/n/$$
for the function $icode f$$.
It is set to the derivative 
$latex \[
\begin{array}{rcl}
dw & = & W_{p-1}^{(1)} ( 0 ) \\
& = &
\partial_u \frac{1}{(p-1) !} \frac{\partial^{p-1}}{\partial t^{p-1}} W(0, 0)
\end{array}
\] $$
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is zero,
all the elements of $icode dw$$ will be instances of $code float$$.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is one,
all the elements of $icode dw$$ will be $code a_float$$ objects.

$head First Order$$
In the case where $latex p = 1$$, we have
$latex \[
\begin{array}{rcl}
dw 
& = & \partial_u \frac{1}{0 !} \frac{\partial^0}{\partial t^0} W(0, 0)
\\
& = & \partial_u W(0, 0)
\\
& = & 
\partial_u \left[ 
w_0 * F_0 ( u + x^{(0)} ) + \cdots +  w_{m-1} F_{m-1} ( u + x^{(0)} )
\right]_{u = 0}
\\
& = & 
w_0 * F_0^{(1)} ( x^{(0)} ) + \cdots + w_{m-1} * F_{m-1}^{(1)} ( x^{(0)} )
\end{array}
\] $$

$head Second Order$$
In the case where $latex p = 2$$, we have
$latex \[
\begin{array}{rcl}
dw 
& = & \partial_u \frac{1}{1 !} \frac{\partial^1}{\partial t^1} W (0, 0)
\\
& = &
\partial_u \left[
	w_0 * F_0^{(1)} ( u + x^{(0)} ) * x^{(1)}
 	+ \cdots + 
	w_{m-1} * F_{m-1}^{(1)} ( u + x^{(0)} ) * x^{(1)}
\right]_{u = 0}
\\
& = &
w_0 * ( x^{(1)} )^\R{T} * F_0^{(2)} ( x^{(0)} ) 
+ \cdots + 
w_{m-1} * ( x^{(1)} )^\R{T} * F_{m-1}^{(2)} ( x^{(0)} )
\end{array}
\] $$

$children%
	example/reverse_1.py%
	example/reverse_2.py
%$$
$head Example$$ 
$table
$rref reverse_1.py$$
$rref reverse_2.py$$
$tend

$end
---------------------------------------------------------------------------
$begin jacobian$$
$spell
	jacobian
	numpy
	adfun
$$

$section Driver for Computing Entire Derivative$$

$index jacobian$$
$index driver, entire derivative$$
$index entire, derivative driver$$
$index derivative, entire driver$$

$head Syntax$$
$icode%J% = %f%.jacobian(%x%)%$$

$head Purpose$$
This routine computes the entire derivative $latex F^{(1)} (x)$$
where $latex F : \B{R}^n \rightarrow \B{R}^m$$ is the 
function corresponding to the $code adfun$$ object $cref/f/adfun/f/$$.

$head f$$
The object $icode f$$ must be an $cref adfun$$ object.
We use $cref/level/adfun/f/level/$$ for the AD $cref ad$$ level of 
this object.

$head x$$
The argument $icode x$$ is a $code numpy.array$$ with one dimension
(i.e., a vector) with length equal to the domain size $cref/n/adfun/f/n/$$
for the function $icode f$$.
It specifies the argument value at which the derivative is computed.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is zero,
all the elements of $icode x$$ must be either $code int$$ or instances
of $code float$$.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is one,
all the elements of $icode x$$ must be $code a_float$$ objects.

$head J$$
The return value $icode J$$ is a $code numpy.array$$ with two dimensions
(i.e., a matrix).
The first dimension (row size) is equal to $cref/m/adfun/f/m/$$
(the number of range components in the function $icode f$$).
The second dimension (column size) is equal to $cref/n/adfun/f/n/$$
(the number of domain components in the function $icode f$$).
It is set to the derivative; i.e.,
$latex \[
	J = F^{(1)} (x)
\] $$
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is zero,
all the elements of $icode J$$ will be instances of $code float$$.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is one,
all the elements of $icode J$$ will be $code a_float$$ objects.

$children%
	example/jacobian.py
%$$
$head Example$$ 
The file $cref jacobian.py$$ contains an example and test of this operation.

$end
---------------------------------------------------------------------------
$begin hessian$$
$spell
	hessian
	numpy
	adfun
$$

$section Driver for Computing Hessian in a Range Direction$$

$index hessian, driver$$
$index driver, hessian$$
$index Lagrangian, hessian$$
$index hessian, Lagrangian$$

$head Syntax$$
$icode%H% = %f%.hessian(%x%, %w%)%$$

$head Purpose$$
This routine computes the Hessian of the weighted sum
$latex \[
	w_0 * F_0 (x) + \cdots + w_{m-1} * F_{m-1} (x)
\] $$
where $latex F : \B{R}^n \rightarrow \B{R}^m$$ is the 
function corresponding to the $code adfun$$ object $cref/f/adfun/f/$$.

$head f$$
The object $icode f$$ must be an $cref adfun$$ object.
We use $cref/level/adfun/f/level/$$ for the AD $cref ad$$ level of 
this object.

$head x$$
The argument $icode x$$ is a $code numpy.array$$ with one dimension
(i.e., a vector) with length equal to the domain size $cref/n/adfun/f/n/$$
for the function $icode f$$.
It specifies the argument value at which the derivative is computed.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is zero,
all the elements of $icode x$$ must be either $code int$$ or instances
of $code float$$.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is one,
all the elements of $icode x$$ must be $code a_float$$ objects.

$head w$$
The argument $icode w$$ is a $code numpy.array$$ with one dimension
(i.e., a vector) with length equal to the range size $cref/m/adfun/f/m/$$
for the function $icode f$$.
It specifies the argument value at which the derivative is computed.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is zero,
all the elements of $icode w$$ must be either $code int$$ or instances
of $code float$$.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is one,
all the elements of $icode w$$ must be $code a_float$$ objects.

$head H$$
The return value $icode H$$ is a $code numpy.array$$ with two dimensions
(i.e., a matrix).
Both its first and second dimension size 
(row and column size) are equal to $cref/n/adfun/f/n/$$
(the number of domain components in the function $icode f$$).
It is set to the Hessian; i.e.,
$latex \[
	H = w_0 * F_0^{(2)} (x) + \cdots + w_{m-1} * F_{m-1}^{(2)} (x)
\] $$
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is zero,
all the elements of $icode H$$ will be instances of $code float$$.
If the AD $cref/level/adfun/f/level/$$ for $icode f$$ is one,
all the elements of $icode H$$ will be $code a_float$$ objects.

$children%
	example/hessian.py
%$$
$head Example$$ 
The file $cref hessian.py$$ contains an example and test of this operation.

$end
---------------------------------------------------------------------------
$begin optimize$$
$spell
	Taylor
	var
$$

$section Optimize an AD Function Object Tape$$

$index optimize$$
$index tape, optimize$$
$index sequence, optimize operations$$
$index operations, optimize sequence$$
$index speed, optimize$$
$index memory, optimize$$

$head Syntax$$
$icode%f%.optimize()%$$


$head Purpose$$
The operation sequence corresponding to an $cref ADFun$$ object can
be very large and involve many operations.
The $icode%f%.optimize%$$ procedure reduces the number of operations,
and thereby the time and memory, required to
compute function and derivative values. 

$head f$$
The object $icode f$$ is an $cref adfun$$ object.

$head Efficiency$$
The $code optimize$$ member function
may greatly reduce the size of the operation sequence corresponding to 
$icode f$$.

$children%
	example/optimize.py
%$$
$head Example$$ 
The file $cref optimize.py$$ contains an example and test of this operation.

$end
---------------------------------------------------------------------------
*/
# include "adfun.hpp"
# include "vector.hpp"
# include "vec2array.hpp"

namespace pycppad {
	// -------------------------------------------------------------

	// constructor for python class ADFun<Base>
	template <class Base>
	ADFun<Base>::ADFun(array& x_array, array& y_array)
	{	vec< CppAD::AD<Base> > x_vec(x_array);
		vec< CppAD::AD<Base> > y_vec(y_array);

		f_.Dependent(x_vec, y_vec);
	}

	// Domain
	template <class Base>
	int ADFun<Base>::Domain(void)
	{	return static_cast<int>( f_.Domain() ); }

	// Range
	template <class Base>
	int ADFun<Base>::Range(void)
	{	return static_cast<int>( f_.Range() ); }

	// Forward
	template <class Base>
	array ADFun<Base>::Forward(int p, array& xp)
	{	size_t    p_sz(p);
		vec<Base> xp_vec(xp);
		vec<Base> result = f_.Forward(p_sz, xp_vec);
		return vec2array(result);
	}

# ifdef NDEBUG
	template <class Base>
	int ADFun<Base>::CompareChange(void)
	{	bool known = true;
		CppAD::ErrorHandler::Call(
			known,
			__LINE__,
			__FILE__,
			"# ifndef NDEBUG",
			"Cannot use CompareChange when NDEBUG is defined"
		);
		return 0;
	}
# else
	// CompareChange
	template <class Base>
	int ADFun<Base>::CompareChange(void)
	{	return static_cast<int>( f_.CompareChange() ); }
# endif

	// Reverse
	template <class Base>
	array ADFun<Base>::Reverse(int p, array& w)
	{	size_t    p_sz(p);
		vec<Base> w_vec(w);
		vec<Base> dw_vec = f_.Reverse(p_sz, w_vec);
		size_t n = f_.Domain();
		vec<Base> result(n);
		for(size_t j = 0; j < n; j++)
			result[j] = dw_vec[j*p + p - 1];
		return vec2array(result);
	}

	// Jacobian
	template <class Base>
	array ADFun<Base>::Jacobian(array& x)
	{	vec<Base> x_vec(x);
		vec<Base> result = f_.Jacobian(x_vec);
		// Kludge: return a vector which is reshaped by cppad.py
		return vec2array(result);
	}

	// Hessian
	template <class Base>
	array ADFun<Base>::Hessian(array& x, array& w)
	{	vec<Base> x_vec(x);
		vec<Base> w_vec(w);
		vec<Base> result = f_.Hessian(x_vec, w_vec);
		// Kludge: return a vector which is reshaped by cppad.py
		return vec2array(result);
	}

	// optimize
	template <class Base>
	void ADFun<Base>::optimize(void)
	{	f_.optimize(); }

	// -------------------------------------------------------------
	// instantiate instances of ADFun<Base>
	template class ADFun<double>;
	template class ADFun<AD_double>;
	// -------------------------------------------------------------

	void adfun_avoid_warning_that_import_array_not_used(void)
	{	import_array(); }
}
